### PR TITLE
kubeadm join: Properly wait for TLS bootstrapping

### LINKED ISSUE #2091: Increase robustness for kubeadm join TLS bootstrap
## Is this a BUG REPORT or FEATURE REQUEST?
BUG REPORT

## Versions

**kubeadm version**: kubeadm v1.17.*

## What happened?
While executing Cluster API tests, in some cases it was observed kubeadm join failures immediately after printing:
```
...
[kubelet-start] Waiting for the kubelet to perform the TLS Bootstrap...
``` 
xref https://github.com/kubernetes-sigs/cluster-api/issues/2769

## What you expected to happen?
To make kubelet-start more resilient by:
1. increasing the TLS bootstrap timeout from to 2 or even 5 minutes (currently 40 sec?).
2. changing the code in order that also annotate CRI is retried as well.

## How to reproduce it (as minimally and precisely as possible)?
This error happens only sometimes, most probably when there is a temporary blackout of the load balancer that sits in front of the API servers (HA proxy reloading his configuration).

## Anything else we need to know?
Important: if possible the change should be kept as small and possible and backported

### PR DESCRIPTION
**Pull request by rosti**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:

The TLS bootstrapping timeout is increased to 5 minutes with a retry once every 5 seconds.
Failing fast if the kubelet is not healthy is also preserved.

**Which issue(s) this PR fixes**:

Fixes kubernetes/kubeadm#2091

**Special notes for your reviewer**:

/cc @kubernetes/sig-cluster-lifecycle-pr-reviews
/area kubeadm
/priority important-soon
/assign @fabriziopandini @neolit123 @ereslibre

**Does this PR introduce a user-facing change?**:

```release-note
kubeadm increased to 5 minutes its timeout for the TLS bootstrapping process to complete upon join
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Review by neolit123 — COMMENTED** [review]
commented on the issue that this is the valid thing to do.
i think the PR needs a release note and we should backport to 1.17 and 18.

/lgtm
/hold

**Comment by neolit123** [comment]
thanks @rosti for debugging and sending the fix.

**Comment by rosti** [comment]
@neolit123 added the release note as requested.

File: cmd/kubeadm/app/cmd/phases/join/kubelet.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/phases/join/kubelet.go:182]: I'm not sure that getting read of WaitForKubeletAndFunc is the right thing to do.
As far as I understand the logic (that is really tangled) is:
- wait for waitForTLSBootstrappedClient to succeed 
or 
- wait for WaitForHealthyKubelet to fail

waitForTLSBootstrappedClient has a 2 minutes timeout
WaitForHealthyKubelet has a 40 sec sleep, and then starts polling with an exponential backoff

TL;DR: we wait for the TLSBootstrapped, but if after some times the kubelet is not healthy, we exit fast

**Review by fabriziopandini — CHANGES_REQUESTED** [review]
@rosti thanks for stepping in on this task!

File: cmd/kubeadm/app/cmd/phases/join/kubelet.go
Comment (neolit123) [reply cmd/kubeadm/app/cmd/phases/join/kubelet.go:182]: if we are keeping WaitForHealthyKubelet this has to be done in the following way:
- don't use WaitForKubeletAndFunc in this PR, potentially remove it because the API is weird. instead run checks serially.
- wait for 40 seconds for the kubelet to start; return on error
- wait for the kubelet.conf to be created - 5 minutes.

Comment (rosti) [reply cmd/kubeadm/app/cmd/phases/join/kubelet.go:182]: I am ok with leaving the healthz check. It's just `WaitForKubeletAndFunc` and the short timeout of `waitForTLSBootstrappedClient` that are problematic ATM.
Certainly, even on a slow running system, the above call to `WaitForKubeletAndFunc` is going to exit after 40s. Hence, we need to ensure that `waitForTLSBootstrappedClient` is completed.

Comment (neolit123) [reply cmd/kubeadm/app/cmd/phases/join/kubelet.go:182]: > Hence, we need to ensure that waitForTLSBootstrappedClient is completed.

like i've mentioned above i'd prefer if we run these checks serially.
`WaitForKubeletAndFunc` is just a weird API.

Comment (rosti) [reply cmd/kubeadm/app/cmd/phases/join/kubelet.go]: I got what @fabriziopandini is saying here and I think he's correct. Hence, I changed the PR to keep the `WaitForKubeletAndFunc` call, but also to have the new 5 minute timeout in `waitForTLSBootstrappedClient`. The polling interval there is also increased to 5 sec.

I do agree that `WaitForKubeletAndFunc` is a weird call. We can certainly replace it with something more appropriate, but let's not do it in this PR if we can.

Comment (neolit123) [reply cmd/kubeadm/app/cmd/phases/join/kubelet.go]: i think we should keep https://github.com/kubernetes/kubeadm/issues/2091 open, until we refactor WaitForKubeletAndFunc.

checking for kubelet.conf during kubelet /healthz errors, feels wrong.

Comment (rosti) [reply cmd/kubeadm/app/cmd/phases/join/kubelet.go]: We probably need a different issue for `WaitForKubeletAndFunc`. After all, if we merge this PR, we would have fixed the original issue there.

**Pull request merged**
